### ISSUE TITLE
garbage collection corrupts registry

### ISSUE THREAD

**Issue by jeffrson** [distribution/distribution#3899]
Hi,
we're running garbage collection monthly for years. Our build scripts daily push changed images, overwriting previous images with the same tag. Usually only a single layer is updated, rarely more (when the base image is updated). We never explicitly delete an image from the registry.

This procedure never caused any problem until last month. Suddenly some images could not be found ("manifest unknown"). I'm not sure, which exact version of registry:2 was installed then, I believe it was 2.8.0. So I updated the image (2.8.1), restarted registry, repushed all missing images and it worked until next gc at the beginning of this month.

Again some images were missing. AFAICT from the logs, gc removed some blob with SHA256 corresponding to "repo digest". Why? It's hard to follow all those SHA256, though. It even seems, like it tried to delete a "repo digest" that (according to the next log) was created later (maybe again?). Strangely, those images (still existing local) have all the same layers, but different "repo digests".

I "fixed" the registry again (could update to another, newer, 2.8.1 (!), restarted, repushed). For testing I run gc every night now. After two or three days the problem reappeared.

I'm at a loss, regarding how to search for the cause and how to fix it. Does something of this description sound familar? Any hint what to do?

Some more observations:
* to fix it, registry must be restarted
* pushing reports (mostly) all layers existing, but the repo digest differs
* I have no idea, why this behaviour started about 1 month ago - nothing changed (afaik)
* From other issues I got the idea to restart registry after gc to clean the cache. Those other issues, however, seem to be always related to deleting a manifest. We don't do this. So why may the "repo digest" be removed, while the image layers still seem to be present so the image could be repushed with no changes in layers, but repo digest only?

**Comment by wy65701436** [comment]
Are you referring to the garbage collection issue on Harbor or Distribution?

If it is Harbor, I'm curious why you had to manually update the registry version. Also, can you explain how you executed garbage collection on Harbor, and what version of Harbor are you currently using?

**Comment by ephrin** [comment]
Had same issue with gc and new layers.
Nightly build upgrades layers of image at 00:00 and later on at 04:00 gc removes some of them. Also those layers can be in multiple images from same subset so multiple manifests (repositories) affected. But usually ONLY those which has upgraded layers.

**Comment by jeffrson** [comment]
After some more research I believe this is related: https://github.com/moby/moby/issues/45509

**Comment by ffdybuster** [comment]
I was also able to observe this multiple times this week with the latest version (2.8.2).

**Comment by domibarton** [comment]
We do have the same issue, even with registry 3!
